### ISSUE TITLE
Unclear error message when running plan without parameters

### ISSUE THREAD

**Issue by joeshannon** [DiamondLightSource/blueapi#206]
When the controller run subcommand is executed without providing any parameters (`-p`), the following is returned:
```
$ blueapi controller run scan
Traceback (most recent call last):
  File "/scratch/30day_tmp/delete15may/bs/blueapi/venv/bin/blueapi", line 8, in <module>
    sys.exit(main())
  File "/scratch/30day_tmp/delete15may/bs/blueapi/venv/lib/python3.10/site-packages/click/core.py", line 1130, in __call__
    return self.main(*args, **kwargs)
  File "/scratch/30day_tmp/delete15may/bs/blueapi/venv/lib/python3.10/site-packages/click/core.py", line 1055, in main
    rv = self.invoke(ctx)
  File "/scratch/30day_tmp/delete15may/bs/blueapi/venv/lib/python3.10/site-packages/click/core.py", line 1657, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/scratch/30day_tmp/delete15may/bs/blueapi/venv/lib/python3.10/site-packages/click/core.py", line 1657, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/scratch/30day_tmp/delete15may/bs/blueapi/venv/lib/python3.10/site-packages/click/core.py", line 1404, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/scratch/30day_tmp/delete15may/bs/blueapi/venv/lib/python3.10/site-packages/click/core.py", line 760, in invoke
    return __callback(*args, **kwargs)
  File "/scratch/30day_tmp/delete15may/bs/blueapi/src/blueapi/cli/cli.py", line 67, in wrapper
    func(*args, **kwargs)
  File "/scratch/30day_tmp/delete15may/bs/blueapi/venv/lib/python3.10/site-packages/click/decorators.py", line 38, in new_func
    return f(get_current_context().obj, *args, **kwargs)
  File "/scratch/30day_tmp/delete15may/bs/blueapi/src/blueapi/cli/cli.py", line 106, in run_plan
    json=json.loads(parameters),
  File "/dls_sw/apps/python/miniforge/4.10.0-0/envs/python3.10/lib/python3.10/json/__init__.py", line 339, in loads
    raise TypeError(f'the JSON object must be str, bytes or bytearray, '
TypeError: the JSON object must be str, bytes or bytearray, not NoneType
```

This is not particularly user friendly, if parameters are always required, can this be made into a mandatory argument?

**Comment by callumforrester** [comment]
This is now a 500 internal server error from a client perspective, which isn't much better

**Comment by callumforrester** [comment]
I believe so, and agreed, and also the CLI should translate HTTP codes into human-readable error messages

**Comment by stan-dot** [comment]
debugging the error in parallel server and client side, 

here the server
![image](https://github.com/DiamondLightSource/blueapi/assets/56644812/b78a8509-af62-4615-b06e-4af749914e02)

here the client
![image](https://github.com/DiamondLightSource/blueapi/assets/56644812/74402860-639e-4410-a771-f2c268209dcc)

**Comment by stan-dot** [comment]
at the moment the cli raises an error only in two cases:

![image](https://github.com/DiamondLightSource/blueapi/assets/56644812/f04cbd55-3131-43c7-93bc-c968828c1c92)

yet the error text is about a `pickle` operation, so a different one from the May 2023 original post. I cannot be sure if this is an issue with setup on my end or just natural changing of the repo

**Comment by tpoliaw** [comment]
The multiprocessing/pickling errors mean it's probably related to #343 if that helps narrow down why it's failing. If you do pass the correct parameters does it run as expected?

**Comment by stan-dot** [comment]
reproducing the error documenting along: 
`name': 'scan',
            'parameter_schema': {'additionalProperties': False,
                                 'properties': {'axes_to_move': {'additionalProperties': {'_axes_to_move': '<class '
                                                                                                           "'bluesky.protocols.Movable'>"},
                                                                 'title': 'Axes '
                                                                          'To '
                                                                          'Move',
                                                                 'type': 'object'},
                                                'detectors': {'items': {'_detectors': '<class '
                                                                                      "'bluesky.protocols.Readable'>"},
                                                              'title': 'Detectors',
                                                              'type': 'array'},
                                                'metadata': {'title': 'Metadata',
                                                             'type': 'object'},
                                                'spec': {'allOf': [{'type': 'string'}],
                                                         'title': 'Spec'}},
                                 'required': ['detectors',
                                              'axes_to_move',
                                              'spec'],
                                 'title': 'scan',
                                 'type': 'object'}},`

**Comment by stan-dot** [comment]
are there plans that do not take any parameters?
`   parameters = parameters or "{}"
    task = RunPlan(name=name, params=json.loads(parameters))`
    `cli.py` lines 184-5. A quickfix would be to reject-return here if no parameters are supplied. 
    
    
    A more complex fix would be to check the list of plans to verify the right number of parameters - which would likely occur on the server.

**Comment by stan-dot** [comment]
also line 185 referenced above uses Task which is referenced in #253

**Comment by tpoliaw** [comment]
> are there plans that do not take any parameters?

There were when `parameters` was changed to be optional and it was annoying that they required passing  `-p '{}'`. It may be that most plans need args now but presumably you can always write ones that don't?

**Comment by stan-dot** [comment]
I'll start with doing this
https://github.com/DiamondLightSource/blueapi/issues/206#issuecomment-1969524593

**Comment by stan-dot** [comment]
added client-side translation of http status codes, addressing @callumforrester 's concern. now working on the server side

**Comment by stan-dot** [comment]
one possible refactor is changing the generic request to a named `requests.get/post` calls that can then be mocked more granularily. it's about this method in `rest.py` 
` def _request_and_deserialize(
        self,
        suffix: str,
        target_type: Type[T],
        data: Optional[Mapping[str, Any]] = None,
        method="GET",
        raise_if: Callable[[requests.Response], bool] = _is_exception,
    ) -> T:
        url = self._url(suffix)
        # todo consider changing this to a switch statement for easier mocking
        # alternatively do single responsiblity principle
        # and decoule request from deserializing.
        response = requests.request(method, url, json=data)
        if raise_if(response):
            message = get_status_message(response.status_code)
            t = response.text
            error_message = f"""Response failed with text: {t},
            with error code: {response.status_code}
            which corresponds to {message}"""
            raise BlueskyRemoteError(error_message)
        deserialized = parse_obj_as(target_type, response.json())
        return deserialized`

**Comment by stan-dot** [comment]
hacky solution the resembles the current tests
`mock_requests.return_value = client.get(f"/plan/{plan}")
# bunch of testing
mock_requests.return_value = client.post("/tasks")
# bunch more testing`

**Comment by stan-dot** [comment]
merely trying to run `blueapi serve` with the config returns an error: `NFO:     Application startup complete.
ERROR:    [Errno -2] Name or service not known
INFO:     Waiting for application shutdown.`

**Comment by stan-dot** [comment]
still got a false negative erro in the test

**Comment by stan-dot** [comment]
`from pydantic import BaseModel, ValidationError, validator
from typing import Optional

class TimeSchema(BaseModel):
    time: float

    # Optional: Add a custom validator if needed
    @validator('time')
    def time_must_be_positive(cls, value):
        if value <= 0:
            raise ValueError('Time must be a positive number')
        return value
`

result - runtime validation won't work  just need to write a separate schema for each plan. which I fully support but that's another ticket. all plans and experiments could be standardized...

I mean we could add manual python validation against the 'properties' but that's not using pydantic as it was supposed to work

**Comment by callumforrester** [comment]
Hmm, yeah, I think separate ticket for custom per-parameter validation

**Comment by stan-dot** [comment]
how to mark this ticket then as pending and dependent on that ticket? I think gitlab has some better support for this kind of thing

**Comment by stan-dot** [comment]
example for future reference: `import json

from pydantic import parse_obj_as
from typing import Any, Optional
from pydantic import BaseModel, Field


class MyBaseModel(BaseModel):
    name = "base"


class MyPlanModel(MyBaseModel):
    name: str = "Plan model"
    description: Optional[str] = Field(
        description="Docstring of the plan", default=None
    )
    parameter_schema: Optional[dict[str, Any]] = Field(
        description="Schema of the plan's parameters",
        alias="schema",
        default_factory=dict,
    )


plan_model: MyPlanModel = MyPlanModel()
plan_model.parameter_schema = {
    "title": "sleep",
    "type": "object",
    "properties": {"time": {"title": "Time", "type": "number"}},
    "required": ["time"],
    "additionalProperties": False,
}


print(plan_model.dict())


wrong_params: str = '{"wrong":"aboslutely wrong"}'


p = json.loads(wrong_params) or "{}"


try:
    validated = parse_obj_as(type(plan_model.parameter_schema), p)
    print(validated)
except ValueError as e:
    print(f"Error: {e}")
    validated = None
`

**Comment by callumforrester** [comment]
From discussion: Validation is already performed server-side and errors are shipped to the client here: https://github.com/DiamondLightSource/blueapi/blob/dba0f257c86f4a3be7d80e9efbb04c463ba52f33/src/blueapi/service/main.py#L152

All the information needed should be in `resp` here: https://github.com/DiamondLightSource/blueapi/blob/dba0f257c86f4a3be7d80e9efbb04c463ba52f33/src/blueapi/cli/cli.py#L187

**Comment by stan-dot** [comment]
todo: 
- [ ] add verbose error messages to the response
